basket/news/views.py

```diff
@@ -119,6 +122,66 @@ def ratelimited(request, e):
     }, 429)
 
 
+def generate_fxa_state():
+    return os.urandom(16).encode('hex')
+
+
+@require_safe
+def fxa_start(request):
+    fxa_oauth = get_fxa_clients()[0]
+    fxa_state = request.session['fxa_state'] = generate_fxa_state()
+    redirect_uri = request.build_absolute_uri('/fxa/success/')
+    url = fxa_oauth.get_redirect_url(state=fxa_state,
+                                     redirect_uri=redirect_uri,
+                                     scope='profile',
+                                     email=request.GET.get('email'))
+    return HttpResponseRedirect(url)
+
+
+@require_safe
+def fxa_success(request):
+    fxa_state = request.session['fxa_state']
+    if not fxa_state:
+        # need a better error state here
+        return HttpResponseJSON({
+            'status': 'error',
+            'message': 'no state found',
+        }, 400)
+
+    code = request.GET.get('code')
+    state = request.GET.get('state')
+    if not (code and state):
+        # need a better error state here
+        return HttpResponseJSON({
+            'status': 'error',
+            'message': 'no code or state returned',
+        }, 400)
+
+    if fxa_state != state:
+        # need a better error state here
```

Comment:
I recommend clearing `request.session['fxa_state']` here if possible, otherwise an attacker gets unlimited guesses at the correct state token (and might even be able to use timing information to reduce the complexity of guessing). It's *probably* fine if you can't conveniently clear it,  because the attack is "I can force you to login to my email preferences page" and that doesn't seem like a worthwhile thing for an attacker to achieve.

Reply:
Ah yes! Good catch. Meant to do that.